server/services/generateReportService.ts

```diff
@@ -14,99 +14,208 @@
-  private readonly log: Logger;
+  logger: Logger;
 
   constructor(
     private readonly initializerContext: PluginInitializerContext,
     esDriver: ElasticsearchPlugin
   ) {
     this.esDriver = esDriver;
-    this.log = this.initializerContext.logger.get();
+    this.logger = this.initializerContext.logger.get();
   }
 
   report = async (req: Request, h: ResponseToolkit): Promise<any> => {
     try {
       const {
         url,
         itemName,
+        source,
         reportFormat,
         windowWidth = 1200,
         windowLength = 800,
       } = req.payload as {
         url: string;
         itemName: string;
+        source: string;
         reportFormat: string;
-        windowWidth: number;
-        windowLength: number;
+        windowWidth?: number;
+        windowLength?: number;
       };
 
       if (reportFormat === FORMAT.png) {
-        const { fileName } = await generatePNG(
+        const { timeCreated, stream, fileName } = await generatePNG(
           url,
           itemName,
           windowWidth,
           windowLength
         );
+        // TODO: temporary, need to change after we figure out the correct date modeling
+        const { callWithRequest } = this.esDriver.getCluster(CLUSTER.DATA);
+        const params: RequestParams.Index = {
+          index: 'report',
+          body: { url, itemName, source, reportFormat, timeCreated },
+        };
+        await callWithRequest(req, 'index', params);
 
-        //@ts-ignore
-        return h.file(`${fileName}.${reportFormat}`, { mode: 'attachment' });
+        return (
+          h
+            .response(stream)
+            .type('image/png')
+            .header('Content-type', 'image/png')
+            //@ts-ignore
+            .header('Content-length', stream.length)
+            .header(
+              'Content-Disposition',
+              `attachment; filename=${fileName}.${reportFormat}`
+            )
+        );
       } else if (reportFormat === FORMAT.pdf) {
-        const { fileName } = await generatePDF(
+        const { timeCreated, stream, fileName } = await generatePDF(
           url,
           itemName,
           windowWidth,
-          windowLength,
-          reportFormat
+          windowLength
+        );
+        // TODO: temporary, need to change after we figure out the correct date modeling
+        const { callWithRequest } = this.esDriver.getCluster(CLUSTER.DATA);
+        const params: RequestParams.Index = {
+          index: 'report',
+          body: { url, itemName, source, reportFormat, timeCreated },
+        };
+        await callWithRequest(req, 'index', params);
+
+        return (
+          h
+            .response(stream)
+            .type('application/pdf')
+            .header('Content-type', 'application/pdf')
+            //@ts-ignore
+            .header('Content-length', stream.length)
+            .header(
+              'Content-Disposition',
+              `attachment; filename=${fileName}.${reportFormat}`
```

Comment:
While streaming will there be any other side effect such as time out  ? size issue ?

Reply:
1. For timeout, do you mean I should set a timeout on server side, such as a socket timeout?
2. For size, it refers to the size of the report pdf/png file(screenshot). I think no matter how large that screenshot will be, we should generate it and deliver to user. Because it's provided by user on the kibana dashboard or kibana visualization tab.

Reply:
Got it. Thanks